### PR TITLE
NMS-10215: Add support for persisting data collection to Kafka

### PR DESCRIPTION
**Pull request by cgorantla**
Build a .proto file for collections (data that match CollectionSet implementations).
Add kafka persister along with kafka producer feature.
Add osgi persister factory as a first step to move persisters to osgi.
Invoke kafka persister when 'osgi' strategy enabled since this is only one registered.
Also Add a way to persist to kafka along with RRD/Newts.

* JIRA: http://issues.opennms.org/browse/NMS-10215

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

File: features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/collection/CollectionSetMapper.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        transactionOperations.execute((TransactionCallback<Void>) status -> {
+            try {
+                OnmsNode node = nodeDao.get(nodeCriteria);
+                nodeResourceBuilder.setNodeId(node.getId());
+                nodeResourceBuilder.setNodeLabel(node.getLabel());
+                nodeResourceBuilder.setForeignId(node.getForeignId());
+                nodeResourceBuilder.setForeignSource(node.getForeignSource());
+                nodeResourceBuilder.setLocation(node.getLocation().getLocationName());
+            } catch (Exception e) {
+               //TODO: Deal with response time resources
```
Comment (j-white) [inline features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/collection/CollectionSetMapper.java]: We should ensure that response time resource are properly handled before merging this.

File: features/collection/persistence-rrd/src/main/java/org/opennms/netmgt/collection/persistence/rrd/RrdPersisterFactory.java [hunk 2]
```diff
@@ -87,5 +102,17 @@ public ResourceStorageDao getResourceStorageDao() {
...
         m_resourceStorageDao = resourceStorageDao;
     }
+
+    private Persister loadKafkaPersister(ServiceParameters params, RrdRepository repository) {
+        Boolean enabled = new Boolean(System.getProperty("org.opennms.timeseries.kafka.persister", "false"));
+        Persister persister = null;
+        if (enabled) {
+            final ServiceLookup serviceLookup = new ServiceLookupBuilder(DefaultServiceRegistry.INSTANCE).blocking()
+                    .build();
+            PersisterFactory persisterFactory = serviceLookup.lookup(PersisterFactory.class, "(strategy=kafka)");
```
Comment (j-white) [inline features/collection/persistence-rrd/src/main/java/org/opennms/netmgt/collection/persistence/rrd/RrdPersisterFactory.java]: I'd prefer to not have to extend the existing persisters with knowledge of another specific persister.

File: features/collection/persistence-osgi/src/main/java/org/opennms/features/collection/persistence/osgi/OsgiPersisterFactory.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    public Persister createPersister(ServiceParameters params, RrdRepository repository) {
+        Persister persister = createPersister(params, repository, false, false, false);
+        return persister;
+    }
+
+    @Override
+    public Persister createPersister(ServiceParameters params, RrdRepository repository, boolean dontPersistCounters,
+            boolean forceStoreByGroup, boolean dontReorderAttributes) {
+        // There is only one persister that can load from osgi, so use the filter directly
+        delegate = SERVICE_LOOKUP.lookup(PersisterFactory.class, "(strategy=kafka)");
```
Comment (j-white) [inline features/collection/persistence-osgi/src/main/java/org/opennms/features/collection/persistence/osgi/OsgiPersisterFactory.java]: We should find a way to avoid hard-coding the "kafka" name here since everything else about this persister is already generic.

File: features/collection/persistence-osgi/src/main/java/org/opennms/features/collection/persistence/osgi/OsgiPersisterFactory.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        }
+        return delegate.createPersister(params, repository);
+    }
+
+    @Override
+    public Persister createPersister(ServiceParameters params, RrdRepository repository, boolean dontPersistCounters,
+            boolean forceStoreByGroup, boolean dontReorderAttributes) {
+        final PersisterFactory delegate = getDelegate();
+        // atleast one of the persister factory should be on osgi registry, else it will use Null Persister.
+        if (delegate == null) {
```
Comment (j-white) [inline features/collection/persistence-osgi/src/main/java/org/opennms/features/collection/persistence/osgi/OsgiPersisterFactory.java]: Can we log something here if we don't find a delegate?

File: features/collection/persistence-rrd/src/main/java/org/opennms/netmgt/collection/persistence/rrd/BasePersister.java [hunk 5]
```diff
@@ -57,6 +59,7 @@
     private final RrdStrategy<?, ?> m_rrdStrategy;
     protected final ResourceStorageDao m_resourceStorageDao;
     private boolean m_dontReorderAttributes = false;
+    private Persister kafkaPersister;
```
Comment (j-white) [inline features/collection/persistence-rrd/src/main/java/org/opennms/netmgt/collection/persistence/rrd/BasePersister.java]: These changes don't appear to be necessary anymore.

File: features/collection/persistence-rrd/src/main/java/org/opennms/netmgt/collection/persistence/rrd/RrdPersisterFactory.java (hunk 2 above)
Comment (j-white) [reply features/collection/persistence-rrd/src/main/java/org/opennms/netmgt/collection/persistence/rrd/RrdPersisterFactory.java]: Can this be removed?

File: features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/collection/KafkaPersisterFactory.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                .getConfiguration(OpennmsKafkaProducer.KAFKA_CLIENT_PID).getProperties();
+        if (properties != null) {
+            final Enumeration<String> keys = properties.keys();
+            while (keys.hasMoreElements()) {
+                final String key = keys.nextElement();
+                producerConfig.put(key, properties.get(key));
+            }
+        }
+        // Overwrite the serializers
+        producerConfig.put("key.serializer", StringSerializer.class.getCanonicalName());
```
Comment (j-white) [inline features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/collection/KafkaPersisterFactory.java]: Can we use the existing constants for "key.serializer" and "value.serializer" here?

File: features/kafka/producer/src/main/resources/OSGI-INF/blueprint/blueprint-kafka-producer.xml [hunk 7]
```diff
@@ -75,10 +75,6 @@
     <property name="alarmSync" value="${alarmSync}"/>
   </bean>
 
-  <bean factory-ref="kafkaProducer" factory-method="setDataSync">
```
Comment (j-white) [inline features/kafka/producer/src/main/resources/OSGI-INF/blueprint/blueprint-kafka-producer.xml]: Do we need to remove this?

File: opennms-services/src/main/java/org/opennms/netmgt/collectd/AliasedResource.java [hunk 8]
```diff
@@ -207,7 +207,7 @@ public String getInstance() {
 
     @Override
     public ResourcePath getParent() {
-        return null; //For node and interface type resources, use the default parent
+        return getCollectionAgent().getStorageResourcePath();
```
Comment (j-white) [inline opennms-services/src/main/java/org/opennms/netmgt/collectd/AliasedResource.java]: Is this change necessary? If so I'd need to take a closer look and see if it has any other impacts.

Comment (cgorantla) [reply opennms-services/src/main/java/org/opennms/netmgt/collectd/AliasedResource.java]: Yeah, this is needed to parse nodeId  for AliasedResource.  
IfInfo does the same way https://github.com/OpenNMS/opennms/blob/develop/opennms-services/src/main/java/org/opennms/netmgt/collectd/IfInfo.java#L273

Comment (j-white) [reply opennms-services/src/main/java/org/opennms/netmgt/collectd/AliasedResource.java]: Sounds good.

File: features/events/daemon/pom.xml [hunk 9]
```diff
@@ -88,6 +88,11 @@
       <scope>test</scope>
       <type>pom</type>
     </dependency>
+    <dependency>
+      <groupId>org.opennms.features.collection</groupId>
+      <artifactId>org.opennms.features.collection.persistence.osgi</artifactId>
```
Comment (j-white) [inline features/events/daemon/pom.xml]: We should add this dependency in `opennms-rrdtool-api`, `opennms-rrd-jrobin` and `org.opennms.features.newts` instead of adding to every module that depends on one of these.

**Comment by cgorantla** [comment]
Looks like bamboo went green.  @j-white Please have another look when you get chance
